pkg/controllers/multicluster/globalservice_controller.go

```diff
@@ -0,0 +1,239 @@
+	ctrllog "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+
+	"github.com/alibaba/hybridnet/pkg/constants"
+	"github.com/alibaba/hybridnet/pkg/controllers/concurrency"
+)
+
+const ControllerGlobalService = "GlobalService"
+
+// GlobalServiceReconciler reconciles a global Service object
+type GlobalServiceReconciler struct {
+	context.Context
+	client.Client
+
+	concurrency.ControllerConcurrency
+}
+
+func (r *GlobalServiceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (result ctrl.Result, err error) {
+	log := ctrllog.FromContext(ctx)
+
+	defer func() {
+		if err != nil {
+			log.Error(err, "reconciliation fails")
+		}
+	}()
+
+	var service = &corev1.Service{}
+	if err = r.Get(ctx, req.NamespacedName, service); err != nil && !apierrors.IsNotFound(err) {
+		return ctrl.Result{}, wrapError("unable to get service", err)
+	} else if apierrors.IsNotFound(err) {
+		// endpoints will be cleaned by owner reference
+		return ctrl.Result{}, nil
+	}
+
+	if !isGlobalService(service) {
+		return ctrl.Result{}, wrapError("unable to unbind remote endpoint slices",
+			r.unbindRemoteEndpointSlices(ctx, service))
+	}
+
+	return ctrl.Result{}, wrapError("unable to bind remote endpoint slices",
+		r.bindRemoteEndpointSlices(ctx, service))
+}
+
+func (r *GlobalServiceReconciler) unbindRemoteEndpointSlices(ctx context.Context, svc *corev1.Service) error {
+	// delete all the endpoint slices created by hybridnet
+	epsList := &discoveryv1beta1.EndpointSliceList{}
+	if err := r.List(ctx, epsList, client.MatchingLabels{
+		discoveryv1beta1.LabelServiceName: svc.Name,
+		discoveryv1beta1.LabelManagedBy:   remoteEndpointSliceControllerName,
+	}, client.InNamespace(svc.Namespace)); err != nil {
+		return fmt.Errorf("unable to list remote endpoint slices for service %v/%v: %v",
+			svc.Name, svc.Namespace, err)
+	}
+
+	for index := range epsList.Items {
+		eps := &epsList.Items[index]
+		if err := client.IgnoreNotFound(r.Delete(ctx, eps)); err != nil {
+			return fmt.Errorf("unable to unbind endpoint slice %v/%v: %v", eps.Name, eps.Namespace, err)
+		}
+	}
+
+	return nil
+}
+
+func (r *GlobalServiceReconciler) bindRemoteEndpointSlices(ctx context.Context, svc *corev1.Service) error {
+	// sync endpoint slices according to remote endpoint slices
+	epsList := &discoveryv1beta1.EndpointSliceList{}
+	if err := r.List(ctx, epsList, client.MatchingLabels{
+		discoveryv1beta1.LabelServiceName: svc.Name,
+		discoveryv1beta1.LabelManagedBy:   remoteEndpointSliceControllerName,
+	}, client.InNamespace(svc.Namespace)); err != nil {
+		return fmt.Errorf("unable to list remote endpoint slices for service %v/%v: %v",
+			svc.Name, svc.Namespace, err)
+	}
+
+	repsList := &multiclusterv1.RemoteEndpointSliceList{}
+	if err := r.List(ctx, repsList, client.MatchingLabels{
+		discoveryv1beta1.LabelServiceName: svc.Name,
```

Comment:
service name is not unique over different namespaces and clusters, do we need more labels as filter here?

Reply:
Another comparison is added, to avoid introducing a new label.

Reply:
ok, even among different namespaces, services with same name are not very common for now